### ISSUE TITLE
What's the intended way of customizing SwaggerUI 3.x?

### ISSUE THREAD

**Issue by awalsh-broadsoft** [swagger-api/swagger-ui#3821]
Is building SwaggerUI with custom files in an npm module the "intended" way to customize SwaggerUI in 3.x?

I'm a total newbie when it comes to customizing SwaggerUI. Swagger's homepage indicates that 3.x should be more customizable, which I find exciting. However, when I try to look into how this is done, I've had trouble finding tutorials or other information.

From the docs inside of the project, it appears as though the "intended" way to customize SwaggerUI is to use it as an npm module, create custom plugins, and make small modifications to existing code.

What is the "intended" way to customize SwaggerUI?

| Q                               | A
| ------------------------------- | -------
| Bug or feature request?         |  
| Which Swagger/OpenAPI version?  | 3.0.0
| Which Swagger-UI version?       | 3.x
| How did you install Swagger-UI? | hosted on Tomcat
| Which browser & version?        |
| Which operating system?         |

**Comment by fgiroud** [comment]
+1 for this one, I need to customize swagger-ui too (not only css but html tags also)    
Is that possible ?

**Comment by legrego** [comment]
My limited understanding is that customizing swagger-ui can be done through plugins, which have the ability to replace existing components by name (e.g., you can provide your own Topbar component to replace the built-in one, if desired). I haven't been able to get this to work though, even following the example [in the FAQ](https://github.com/swagger-api/swagger-ui/wiki/FAQ#how-to-hide-topbar).

The other recommendation I've seen is to fork the repo, and provide your own [StandaloneLayout](https://github.com/swagger-api/swagger-ui/blob/170aa1234924d189bbb65b6f839fb4001412e85b/src/standalone/layout.jsx#L4). At that point though, you can make all the code changes you want, which IMO probably isn't the preferred way to customize your instance.

tl;dr: +1

**Comment by webron** [comment]
While still WIP, take a look at https://github.com/swagger-api/swagger-ui/tree/ft/docs/docs.

**Comment by shockey** [comment]
^ the docs are a good starting place, and are mostly stable. There's an inconsistency with wrapComponent that I still need to dig into, which is why it isn't in master yet.

From a more general perspective - if you'd describe the changes you want to make to Swagger-UI like "I want to...":
- "modify the overall layout"
- "hide [thing]"
- "hide [thing] based on certain conditions"
- "inject this [HTML || React component] before/after [thing]"
- "mutate the data going into [thing]"
- "run some custom code that causes side effects when [event] happens"
- "modify the data that flows into the application when [event] happens"
- "disable the default behavior for [event]"

... then the plugin system will most likely support your use case. In fact, the entire Swagger-UI core application is based on a collection of built in plugins that you can override with your own.

There are limitations, though - for example, some parts of Swagger-UI use rather complex components that nest a lot of HTML into one place, which is hard to override. As an example, until recently you'd have to reimplement the entire Model component as well. A user popped up wanting to modify individual properties in a model, so they ended up opening a PR (#3868) that creates a Property component that allows for granular control of how properties are rendered from a plugin.

I'd suggest starting by trying to accomplish what you want via a plugin. If you ever outgrow that (most folks won't), you can take your existing plugin code with you into a fork of Swagger-UI.

> From the docs inside of the project, it appears as though the "intended" way to customize SwaggerUI is to use it as an npm module, create custom plugins, and make small modifications to existing code.
> 
> What is the "intended" way to customize SwaggerUI?

Basically, yes: I suggest using Swagger-UI as a module and consuming it in a codebase with its own build system, so you can take advantage of modern JS syntax and JSX. I have an [example of such a thing, `swagger-ui-cornify`](https://github.com/shockey/swagger-ui-cornify) that you can use as a starting point.

---

CSS is a bit trickier, since it's not part of the plugin system, or JS-land at all. The classic approach of just overriding classes should work for now, but be aware that CSS is not part of our public API, so things may change in an incompatible way between versions - there's no real way around that, apart from bumping our major version whenever we rename a class, which is just silly.

If the plugin system is giving you trouble (cc @legrego) feel free to open an issue, and we'll mark it as `Support` and get you unstuck.

**Comment by thompsongl** [comment]
Here's the plugin approach we've taken (re: #3868):
https://gist.github.com/thompsongl/bd11ef33b7ae86e1841990ba8587ce4c

First, added a more focused component (`Property` in #3861). Then, in our project added custom components (e.g., `EnumTable`) registered as a plugins. Those components are consumed by a `wrapComponents` plugin (wrapping `Property`), which is registered as a preset (required if wrapping or replacing a core component). That wrapped component then conditionally renders one of our custom components or the original component depending on the presence of certain spec `x-` extensions.

As @shockey mentioned, using `wrapComponents` with complex/HTML-heavy components is difficult. My guess is that most of the complex core components will naturally be converted to composing single-focus, presentational components as more people attempt to customize. Luckily, React makes this very easy. I've found that React plus access to `system`/`toolbox` in the creation fn makes for pretty smooth customization.

**Comment by shockey** [comment]
Closing due to inactivity. 

This is simply to keep our issue tracker clean - feel free to comment if there are any further thoughts or concerns, and we'll be happy to reopen this issue.
